### DISCUSSION TITLE
OrmLiteValidationSource can not work well when using redisCache because of  Json Deserialize

### DISCUSSION THREAD

**Discussion by 1257960069** [ServiceStack/Discuss#108] [Q&A] [answered]
```
public class ConfigureRedis : IHostingStartup
{
    public void Configure(IWebHostBuilder builder) => builder
        .ConfigureServices((context, services) => {
            services.AddSingleton(c => c.GetRequiredService<IRedisClientsManager>().GetCacheClient());
            services.AddSingleton<IRedisClientsManager>(new RedisManagerPool(context.Configuration.GetConnectionString("Redis")));
        })
        .ConfigureAppHost(appHost => {
            // Enable built-in Redis Admin UI at /admin-ui/redis
            // appHost.Plugins.Add(new AdminRedisFeature());
        });
}
```

```
public class ConfigureValidation : IHostingStartup
{
    // Add support for dynamically generated db rules
    public void Configure(IWebHostBuilder builder) => builder
        .ConfigureServices(services =>
        {
            services.AddSingleton<IValidationSource>(c =>
            {
                var cache = c.GetRequiredService<ICacheClient>();
                return new OrmLiteValidationSource(c.GetRequiredService<IDbConnectionFactory>(), cache, TimeSpan.FromMinutes(5));
            });
        })
        .ConfigureAppHost(appHost =>
        {
            appHost.Resolve<IValidationSource>().InitSchema();
        });
}
```

the fix a way is 
```JsConfig.AllowRuntimeType = _ => true;```

isssue code

```
public class OrmLiteValidationSource : IValidationSource, IRequiresSchema, IValidationSourceAdmin, IClearable
{
    public IEnumerable<KeyValuePair<string, IValidateRule>> GetValidationRules(Type type)
    {
        if (Cache == null)
        {
            return GetDbValidationRules(type);
        }

        if (!CacheDuration.HasValue)
        {
            return Cache.GetOrCreate("IValidationSource." + type.Name, () => GetDbValidationRules(type));
        }

        return Cache.GetOrCreate("IValidationSource." + type.Name, CacheDuration.Value, () => GetDbValidationRules(type));
    }
 //issue: this return type not can DeserializeObject by default
    private IEnumerable<KeyValuePair<string, IValidateRule>> GetDbValidationRules(Type type)
    {
        using IDbConnection dbConn = OpenDbConnection();
        return dbConn.Select((ValidationRule x) => x.Type == type.Name && x.SuspendedDate == null).Map((ValidationRule x) => new KeyValuePair<string, IValidateRule>(x.Field, x));
    }

```

**Comment by mythz** [comment]
Use `MemoryCacheClient` or `ServiceStackHost.DefaultCache` instead:

```csharp
services.AddSingleton<IValidationSource>(c => {
    var cache = c.GetRequiredService<MemoryCacheClient>();
    return new OrmLiteValidationSource(c.GetRequiredService<IDbConnectionFactory>(), cache, TimeSpan.FromMinutes(5));
});
```

**Reply by 1257960069** [reply]
why not can use redis?

**Reply by mythz** [reply, marked as answer]
It's for locally cached validation rules that rarely change, I'd recommend avoiding the the unnecessary I/O.

**Reply by 1257960069** [reply]
I suggest removing the construction parameter with ICacheClient cache=null

**Reply by mythz** [reply]
Then it would hit DB each time.

**Reply by 1257960069** [reply]
What I mean is that internal memory caching is used, and custom caching is not allowed

**Reply by mythz** [reply]
Not in the habit of adding breaking changes that lose flexibility, users should be able to configure it with how they want.

[OrmLiteValidationSource.cs](https://github.com/ServiceStack/ServiceStack/blob/main/ServiceStack/src/ServiceStack.Server/OrmLiteValidationSource.cs) is also a fairly simple class should you want to locally copy & modify.

**Reply by 1257960069** [reply]
ok,thanks.  by the way. I need a role and permission system.  but servicestak just use can roles that declared roles from dto.  otherwise, RequiresAnyPermissionAttribute and RequiresAnyPermissionAttribute can not work at mvc controller.  I think servicestack should can enhance the  mvc, rather than as as single framework not can work with mvc

**Reply by mythz** [reply]
You should submit any feature requests to [servicestack.net/ideas](https://servicestack.net/ideas) instead, although our latest ASP.NET Core templates use Identity Auth you should use the standard `[Authorize]` attributes instead.

>  I think servicestack should can enhance the mvc, rather than as as single framework not can work with mvc

No idea what this means, our latest .NET 8 MVC and Razor Pages templates https://servicestack.net/start happily work side-by-side with MVC, Razor Pages and Blazor which now all use Endpoint Routing + ASP.NET Core IOC + Identity Auth so they should work together nicely. We don't intend to "Enhance MVC", just ensure that ServiceStack APIs work well when configured within the same ASP.NET Core App so you can use MVC "untouched" as intended.

**Reply by 1257960069** [reply]
thanks

**Reply by 1257960069** [reply]
ASP.NET Core Identity  was not offer  a ui to manage user and role. just has some manage class for calling. ids has not ui.  servicestack has admin-ui/users,  but can not  add a role and permission dynamic. May be I need a framework like abp. servicestack  do something ,but  not do all,  right?

**Reply by mythz** [reply]
I'm having a hard time trying to understand your english, but you'll need to use ASP.NET Core's Identity classes to manage roles, e.g. checkout how we initially Create Roles, Users and adding Roles to users:

https://github.com/NetCoreTemplates/blazor/blob/main/MyApp/Configure.Db.Migrations.cs#L47

Please open up new Threads if you want to ask new questions that are unrelated to the Thread Topic.

**Reply by 1257960069** [reply]
thanks，bye，not new question
